### ISSUE TITLE
Input clears itself when date is invalid

### ISSUE THREAD

**Issue by uvtzxpm** [gpbl/react-day-picker#736]
If I try to change a date from `July 25, 2018` to `July 25, 2020` but while doing so I change the date to something invalid, such as `July 25, 202`, the input clears itself.

So, I have to be very careful not to let the date become invalid. It's a bit frustrating for my users.

I've made a GIF to show what I mean:
![demo](https://i.imgur.com/oMiOnoQ.gif)

(using this page: http://react-day-picker.js.org/examples/input-from-to/)

I'm not sure if this is considered a "bug", or if I should be doing something different with the parsing myself. If it's not a bug, do you have any hints on how to work around it?

Other demos don't have the problem, such as:
http://react-day-picker.js.org/examples/input-moment/
http://react-day-picker.js.org/examples/input-state/

Maybe it's the combination of moment parsing and React state?

**Comment by gpbl** [comment]
Thanks @uvtzxpm I've looked a bit deeper into this and the bug has been introduced in #647 to fix #614.

(my investigation follows :) 

Let's take this example: http://react-day-picker.js.org/examples/input-state/

```jsx
handleDayChange(selectedDay, modifiers) {
    this.setState({
      selectedDay,
      isDisabled: modifiers.disabled === true,
    });
}
```

When the user types a day that is not a valid date, we decided in [#647](https://github.com/gpbl/react-day-picker/blob/master/src/DayPickerInput.js#L407-L410) to call `onDayChange` with `undefined`:

```js
if (!day) {
  this.setState({ value });
  if (onDayChange) onDayChange(undefined, {});
  return;
}
```

In the example above, we will set `selectedDay` to `undefined`, pass `undefined` to the `value` prop and the component will fall into the prop's default value – which is an empty string.

We have to understand how it should behave instead. 
- we want the input to display the typed value (even if it is an invalid date)
- we want `onDayChange` being called with `undefined` and the currently selected day to be `undefined`.

I need to think about it.  My first thoughts are to play with a `dateIsInvalid` or `rawValue` in the component state.
